### DISCUSSION TITLE
Handling Whitespace Only Usernames

### DISCUSSION THREAD

**Discussion by Dan-613** [jellyfin/jellyfin#12967] [Q&A]
I was wondering what your thoughts are on the best way to handle this issue.

I'm looking at this issue.
https://github.com/jellyfin/jellyfin-web/issues/6253
Exception thrown when the username of ' ' is used during login.

And I'm trying to work out how best to contribute.

The secondary issue mentioned in discussion on that bug report is that there could be legacy user called ' ' which would need to be migrated. I'm not sure whether that is an over stretch since that user would not be able to log in due to the exception.

And finally, if a user accidentally includes a whitespace in their username during creation or login, login is not possible https://github.com/jellyfin/jellyfin/issues/8897

For Consideration...  There could legitimately be intentional whitespaces at the beginning or end of usernames so we can't just trim() everything. Especially considering that there could be a user called 'Bob' and a user called 'Bob '.

From what I can tell, the main bug can be handled in a couple of ways, but I'd like to discuss which way the community thinks is the most suitable approach.

1. Implement a migration which automatically renames ' ' users to User + random number. eg. User001. The server admin would then have to log in to see why the ' ' user is no longer able to log in and discover the new username.
2. Open a dialog when the admin logs into the web interface which prompts admin to rename the user ' ' to something else.

Once one of these can be implemented, we add code to the server-side login to protect against the exception caused when the username supplied is ' ', and prevent the creation of new users called ' ' (assuming that protection doesn't already exist).

With regards to the 2nd bug report, where a whitespace is accidentally included at the end of a username, we could attempt logging in with the whitespace, and if that fails, we could .trim() and then attempt login again before returning a result. But that seems kind of messy to me.

What is the community's thoughts on this? It's the first time I'm looking to contribute to an opensource project and I'd like some guidance so I get it right.

**Comment by TheMelmacian** [comment]
Imho usernames with leading or trailing whitespaces make much more hassle than giving any benefit. I would forbid them completely going forward. Can be checked via regex, or a length comparison before and after trim, with a corresponding error message. If needed the user could use an underscore. So 'Bob ' is not allowed, but 'Bob_'.
If such users already exist I would leave them as is to not break existing installations. An admin has to adapt them manually if needed.

If someone really wants to use such usernames, he has to create/edit them directly in the database but with no official support.

**Reply by Dan-613** [reply]
So you would suggest preventing further accounts from containing a space in their usernames.

> If such users already exist I would leave them as is to not break existing installations

How would you achieve this while also trimming any accidental whitespaces that people include in the login screen?

**Reply by TheMelmacian** [reply]
A space in the middle is OK ('Bob Bob' etc.), but I would prevent new accounts with leading or trailing whitespaces.

If you want to trim all usernames on login, affected users stop working. So you have to migrate them in some way, which can be problematic if two users only differ in whitespaces. I assume not many installs exists where this is the case but it's theoretically possible.
It's also possible to leave them as they are and inform the admins that they have to adapt the usernames manually.

An alternative would be to not trim on login but warn the user in some way if any leading or trailing whitespaces are detected, or highlight the whitespace so the user sees that he made a mistake.

**Reply by Dan-613** [reply]
> If you want to trim all usernames on login, affected users stop working. So you have to migrate them in some way, which can be problematic if two users only differ in whitespaces. I assume not many installs exists where this is the case but it's theoretically possible.

We could trim all and then in the unlikely event that there are users that only differ in trailing whitespace, we prevent login with a message saying "Two users found with similar/matching usernames. Please ask admin to rename the accounts."

**Reply by TheMelmacian** [reply]
If we change existing users there is always a risk of ending up with duplicate names.
So if we decide that existing users should be migrated, we always need to find a solution for the edge cases.
This is probably one of the cleanest ways, because most users will end up correctly.

The only problem is when an admin account is affected. In this case, the login to the admin account must get priority and must not be blocked.
If multiple admin accounts end up with the same name we need to decide which one gets priority: the oldest, the one with the latest successful login, etc.?

**Comment by SwatterMatter** [comment]
I am not  a Jellyfin developer, but I think a possible solution that would be backwards compatible would be to trim leading and trailing whitespaces from usernames during login only if the username was not found. The system checks for the username as is, then, if no Username was found, it can check to see if the submitted username has trailing or leading spaces. If it does, strip them and check again. This would allow logins with an accidental whitespace to get in and would not require the usernames on the server be changed. This would solve the current problem while giving the programmers time to figure out how to migrate the problematic usernames.
